pkg/registry/apis/provisioning/repository_controller.go

```diff
@@ -161,21 +164,38 @@ func (rc *RepositoryController) deleteRepository(obj interface{}) {
 
 // processNextWorkItem deals with one key off the queue.
 // It returns false when it's time to quit.
-func (rc *RepositoryController) processNextWorkItem(_ context.Context) bool {
+func (rc *RepositoryController) processNextWorkItem(ctx context.Context) bool {
 	item, quit := rc.queue.Get()
 	if quit {
 		return false
 	}
 	defer rc.queue.Done(item)
 
-	rc.logger.Info("RepositoryController processing key", "key", item)
+	logger := rc.logger.With("key", item.key)
+	logger.InfoContext(ctx, "RepositoryController processing key")
 
 	err := rc.processFn(item)
 	if err == nil {
 		rc.queue.Forget(item)
 		return true
 	}
 
+	item.attempts++
+	logger = logger.With("error", err, "attempts", item.attempts)
+	if item.attempts >= maxAttempts {
+		logger.ErrorContext(ctx, "RepositoryController failed too many times")
+		rc.queue.Forget(item)
+		return true
+	}
+
+	if !apierrors.IsServiceUnavailable(err) {
+		logger.InfoContext(ctx, "RepositoryController will not retry")
+		rc.queue.Forget(item)
+		return true
+	} else {
+		logger.InfoContext(ctx, "RepositoryController will retry as service is unavailable")
+	}
```

Comment:
Shouldn't we _always_ retry on ServiceUnavailable and HTTP 429 (whatever the Go constant is)?

Reply:
do you mean that we should not have max retries at all? @Proximyst

Reply:
I'm not sure. I reckon for errors that aren't one of those, we should probably try another time or two, but we should ideally juts make it work when stuff fails due to (highly likely) transient errors.

Reply:
Probably for that we should introduce some exponential backoff. I have parked this into this other one: https://github.com/grafana/git-ui-sync-project/issues/49

For now this one will help us avoid too many retries.